compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/util/ImmutableTypeCloner.java

```diff
@@ -413,16 +413,22 @@ private static BIntersectionType setImmutableType(Location pos, Types types,
             default:
                 BUnionType origUnionType = (BUnionType) type;
 
-                BType immutableEffectiveType;
                 LinkedHashSet<BType> originalMemberList = origUnionType.getMemberTypes();
-                LinkedHashSet<BType> readOnlyMemTypes = new LinkedHashSet<>(originalMemberList.size());
-                immutableEffectiveType = BUnionType.create(origUnionType.tsymbol);
+                if (unresolvedTypes.contains(origUnionType) && origUnionType.immutableType != null) {
+                    return origUnionType.immutableType;
+                } else {
+                    BIntersectionType immutableUnionIntersectionType = createImmutableIntersectionType(env,
+                            origUnionType, BUnionType.create(origUnionType.tsymbol), symTable);
+                    origUnionType.immutableType = immutableUnionIntersectionType;
```

Comment:
```suggestion
                    origUnionType.immutableType = createImmutableIntersectionType(env,
                            origUnionType, BUnionType.create(origUnionType.tsymbol), symTable);
```

Reply:
We don't need the else block right? We can move the logic out since we're returning in the if block?

Reply:
Remove else condition